File: python/1337.the-k-weakest-rows-in-a-matrix.py
from heapq import heapify, nsmallest


def get_zero(row: list) -> int:
    lo, hi = 0, len(row)
    while lo < hi:
        mid = (lo + hi) // 2
        if row[mid] == 0:
            hi = mid
        else:
            lo = mid + 1
    return lo


class Solution:
    def kWeakestRows(self, mat: List[List[int]], k: int) -> List[int]:
        res = []
        for i, row in enumerate(mat):
            res.append((get_zero(row), i))

        heapify(res)
        print(res)
        res = nsmallest(k, res)
        return list(map(lambda x: x[1], res))
        

File: cpp/1337.the-k-weakest-rows-in-a-matrix.cpp
class Solution {
private:
    typedef pair<int, int> pi;
    
    int getZero(const vector<int>& row) {
        int lo = 0;
        int hi = row.size();

        while (lo < hi) {
            int mid = (lo + hi) / 2;
            if (row[mid] == 0) {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }

        return lo;
    }

public:
    vector<int> kWeakestRows(vector<vector<int>>& mat, int k) {
        priority_queue<pi, vector<pi>, greater<pi>> pq;
        int j = 0;
        for (const auto& row : mat) {
            pq.push({getZero(row), j});
            j++;
        }

        vector<int> res;

        for ( ; k > 0; k--) {
            res.push_back(pq.top().second);
            pq.pop();
        }

        return res;
    }
};